expression/builtin_grouping.go

```diff
@@ -0,0 +1,206 @@
+	"github.com/pingcap/tidb/util/chunk"
+	"github.com/pingcap/tipb/go-tipb"
+)
+
+var (
+	_ functionClass = &groupingImplFunctionClass{}
+)
+
+var (
+	_ builtinFunc = &builtinGroupingImplSig{}
+)
+
+type groupingImplFunctionClass struct {
+	baseFunctionClass
+}
+
+func (c *groupingImplFunctionClass) getFunction(ctx sessionctx.Context, args []Expression) (builtinFunc, error) {
+	if err := c.verifyArgs(args); err != nil {
+		return nil, err
+	}
+	argTp := []types.EvalType{types.ETInt}
+	bf, err := newBaseBuiltinFuncWithTp(ctx, c.funcName, args, types.ETInt, argTp...)
+	if err != nil {
+		return nil, err
+	}
+	bf.tp.SetFlen(1)
+	sig := &builtinGroupingImplSig{bf, 0, map[int64]struct{}{}}
+	sig.setPbCode(tipb.ScalarFuncSig_GroupingSig)
+	return sig, nil
+}
+
+// grouping functions called by user is actually executed by this builtinGroupingImplSig.
+// Users will designate a column as parameter to pass into the grouping function, and tidb
+// will rewrite it to convert the parameter to the meta info. Then, tidb will generate grouping id
+// which is a indicator to be calculated with meta info, these grouping id are actually what
+// builtinGroupingImplSig receives.
+type builtinGroupingImplSig struct {
+	baseBuiltinFunc
+
+	// TODO these are two temporary fields for tests
+	mode          tipb.GroupingMode
+	groupingMarks map[int64]struct{}
+}
+
+func (b *builtinGroupingImplSig) SetGroupingMode(mode tipb.GroupingMode) {
+	b.mode = mode
+}
+
+func (b *builtinGroupingImplSig) SetMetaGroupingMarks(groupingMarks map[int64]struct{}) {
+	b.groupingMarks = groupingMarks
+}
+
+func (b *builtinGroupingImplSig) getGroupingMode() tipb.GroupingMode {
+	return b.mode
+}
+
+// metadata returns the metadata of grouping functions
+func (b *builtinGroupingImplSig) metadata() proto.Message {
+	args := &tipb.GroupingFunctionMetadata{}
+	*(args.Mode) = uint32(b.mode)
+	for groupingMark, _ := range b.groupingMarks {
+		args.GroupingMarks = append(args.GroupingMarks, uint64(groupingMark))
+	}
+	return args
+}
+
+func (b *builtinGroupingImplSig) Clone() builtinFunc {
+	newSig := &builtinGroupingImplSig{}
+	newSig.cloneFrom(&b.baseBuiltinFunc)
+	newSig.mode = b.mode
+	newSig.groupingMarks = b.groupingMarks
+	return newSig
+}
+
+func (b *builtinGroupingImplSig) getMetaGroupingMarks() map[int64]struct{} {
+	return b.groupingMarks
+}
+
+func (b *builtinGroupingImplSig) getMetaGroupingID() int64 {
+	var metaGroupingID int64
+	grouping_ids := b.getMetaGroupingMarks()
+	for key := range grouping_ids {
+		metaGroupingID = key
+	}
+	return metaGroupingID
+}
+
+func (b *builtinGroupingImplSig) checkMetadata() error {
+	mode := b.getGroupingMode()
+	grouping_ids := b.getMetaGroupingMarks()
+	if mode != tipb.GroupingMode_ModeBitAnd && mode != tipb.GroupingMode_ModeNumericCmp && mode != tipb.GroupingMode_ModeNumericSet {
+		return errors.Errorf("Mode of meta data in grouping function is invalid. input mode: %d", mode)
+	} else if (mode == 1 || mode == 2) && len(grouping_ids) != 1 {
```

Comment:
why still use `mode == 1 || mode == 2` here

Reply:
> why still use `mode == 1 || mode == 2` here

I will change it